### ISSUE TITLE
Multi-steplib zis proposal

### ISSUE THREAD

**Issue by 1000TurquoisePogs** [zowe/zss#711]
As requested in https://github.com/zowe/zss/issues/703, this ticket is to discuss an implementation proposal

Goal: have `zwe components install -o /the/extension[.pax]` go beyond its current task of just copying unix content into the extensions directory, but also to update the ZIS STC steplib list.

# Idea 1
`zwe components install -o /the/extension[.pax] -d THE.DS1,THE.DS2`
... Have -d serve as a list of datasets to append to the steplib.

Problems: 
* `zwe components upgrade` and `zwe components uninstall` exist. How do we know which steplibs to REMOVE? How can we track which extension added which steplib entry? a file in the workspace directory? something in zowe.yaml?

* I don't like this being manual. How can we detect the datasets so that users do not need to be prompted, and cannot do something wrong? I'd like to query SMPE's knowledge somehow.

# Idea 2
No new -d parameter, just manifest info & SMPE lookup.
... How can I query what datasets on a system are associated with an FMID? I think that's superior. If a Zowe extension manifest could have a field for stating its FMID, like:

```yaml
name: angular-sample
id: org.zowe.zlux.sample.angular
version: 2.16.0
fmid: AZWE999
```

Then, can I get a list of datasets associated with it via something involving https://www.ibm.com/docs/en/zos/3.1.0?topic=command-listing-by-fmid-fmidset ?

And, if there were multiple datasets, how do I know which one goes in the steplib?
Would we need a suffix pattern to follow, such as:

```yaml
name: angular-sample
id: org.zowe.zlux.sample.angular
version: 2.16.0
fmid: AZWE999 # Proposed new attribute. Used with steplibLLQ to find exact dataset.
zisPlugins:
  - id: plugin1
    steplibLLQ: ZISPLUG1 # Proposed new pattern. No copy occurs, but steplib entry added
  - id: plugin2
    steplibLLQ: ZISPLUG2
  - id: plugin3
    path: "/zisServer" # Current zowe pattern. it copies from here to the zis plugin ds
```

**Comment by JoeNemo** [comment]
Idea 1 needs to gather everything that is currently needed from Zowe base and all components (if i understand correctly) and remove what is no longer needed and add STEPLIB entries for everything that is needed.   While not impossible, this sounds pretty complex.

**Comment by JoeNemo** [comment]
In Idea 2, what is the requirement that SMP/E actually know all STEPLIBs?   Did we state this as a requirement for component developers?

**Comment by 1000TurquoisePogs** [comment]
> Idea 1 needs to gather everything that is currently needed from Zowe base and all components (if i understand correctly) and remove what is no longer needed and add STEPLIB entries for everything that is needed. While not impossible, this sounds pretty complex.

I know how to write it, but it sounds accident-prone. Users will just type the wrong thing and erase or add the wrong steplib.
If there was some eyecatcher in a zis plugin it would help. but is there?

**Comment by P-E-J** [comment]
Great to see you want to keep it automated but a STEPLIB change by hand is very common. So if I need to install an extra plugin and the manual state 'add this library to the ZWESLSTC steplib' it will work perfectly. I would make it clearer to me, especially while I have to do this anyway in our version control program in a place where we keep all proclibs together.  From this central place we have a secure pipeline to distrubute all stc to all our sysplexes (10x).

**Comment by 1000TurquoisePogs** [comment]
@P-E-J about automation:

Although the discussion was originally about datasets, the command `zwe components install` today also includes unix ZFS operations.
 
To avoid running that command at all, you would need to copy or symbolic link an extension's ZFS content into the "zowe.extensionDir" directory manually, and ensure one of your YAML lists the extension as "enabled: true". Some products seem to ship with a YAML that says this and request users to append that YAML to their YAML config list, but the ZFS directory copy has so far been automated. Is that a good enough solution for you to perform install/upgrade manually or do you have other ideas?

**Comment by nosrednayduj** [comment]
We have some rexx code, called from a shell script, that adds STEPLIB statements to AUX task proclib entries. This is called separately from any installation process. So that allows the user to choose whether or not to use a different process. The shell script makes a temporary file and if it doesn't have access to actually change proclib (frequent), then the user can use the temporary file to help with whatever their normal process is. So I don't really want you to add to STEPLIB in zwe component install; I just want to have the option of supplying those libraries in a data set already and then adding that data set to authPluginLib. If you want to make a separate function to add the STEPLIB, that's fine. Perhaps I can supply the rexx.

**Comment by Gautham-coder** [comment]
Hi,
I have implemented a prototype with IDEA 1 where we supply the the list of dataset through arguement '-d', as shown below:

zwe components install -o /u/XXXX/file-manager-1.3.0-20241112.201716.pax -c example-zowe.yaml **-d DSN1, DSN2**...

Solution:
======

- I process the input list to check for valid DSNs:
       1.  check for duplicates,
       2.  wrong dataset names/length and
       3.  if dataset exists.

- When adding the dataset to the STEPLIB list:
      1.  Check if the dataset exits in the list if so don't add.
      2.  if commented then uncomment it.
      3.  Also if there a empty lines or comments in the JCL to skip them correctly
     
Suggestion needed:
=============
I need help on if there are any other scenarios in the steplib section of JCL that needs to be taken care of.

Before:
![STEPLIB_Before](https://github.com/user-attachments/assets/5996d10c-7f84-4b03-b790-11420086c484)


After:
![STEPLIB_After](https://github.com/user-attachments/assets/02ad4abc-b494-4c30-9303-bd7da46ade6f)

**Comment by nosrednayduj** [comment]
So we have a customer who wants this feature. However what they really want is the "do not copy the executables to the authPluginLib" part, and not so much the "update STEPLIB" part. Is there any need for the data set names other than updating STEPLIB? That is, can we have an option to say "install component without copying plug-in libraries" and "I promise to be good and put the STEPLIB in myself"?

**Comment by Gautham-coder** [comment]
The implementation is such that when a user supplies the plug-in datasets with argument '-d', we update the steplib entries as well as skip copying the zis-plugin into the apf plugin dataset.

**Comment by nosrednayduj** [comment]
As I mentioned in a direct message conversation, the customer wants "hands off my data sets", and promises to do STEPLIB updates manually, so we need another option to "don't update STEPLIB"

**Comment by 1000TurquoisePogs** [comment]
If the user wants no actions taken, then potentially no code needs to be written.
It could be done via documentation.
But otherwise, it sounds like a `--dry-run` mode.

**Comment by nosrednayduj** [comment]
From the user perspective, zwe components install would need to update the extensions soft link. Are there other things besides creating the soft link and copying the ZIS plug-ins that components install does? I think we need those done too; it's only the data sets that he wants to leave untouched. So it's not quite "dry".

We might want to talk about what the output of dry run with look like, because if UMS is calling it, it will be confusing to the user to have it say "would copy these files to this data set" for each plug-in, when they don't actually care about the location of plug-in files on UNIX, but for backwards compatibility we would have to ship the plug-in libraries in both UNIX and new data set locations, because if you are upgrading you don't want to have both the new location and the authPluginLib in your STEPLIB because what if you put them in the wrong order and now you have the old code that was in authPluginLib and the new code that's in the shipping in new place.

**Comment by 1000TurquoisePogs** [comment]
I would like to avoid creating half-automated half-manual activities that only pertain to 1 user at the expense of adding confusion & user error opportunities for other users, and maintenance/test burden for us.
So I'd rather either document the tasks that can be done manually, and/or have a user to run sub-commands.

`zwe components install` is documented to have subcommands. Already that creates such opportunities for error and confusion but they already exist and are documented so they're not going away.

It operates on the subcommands like so

1) runs `zwe components install extract` 
* ensures the zowe.extensionDirectory exists
* checks if the input is a directory or a file archive
    * if directory, creates a temporary symlink in zowe.extensionDirectory
    * if file archive, tries to extract it into a temporary folder in zowe.extensionDirectory
* verifies that the result has a valid manifest against schema, to avoid user error or corruption
* if the manifest of the component requests the component to be auto-tagged upon install, runs the tag-files.sh script to tag files according to their filename extensions
* checks for and verifies if the extension has any plugins, to avoid developer error or corruption
* renames the temporary directory to the name of the component as seen in the manifest, indicating successful extract

2) runs `zwe components install process-hook`
* if the extension has an install script, runs it
* checks every appfw plugin in the extension for zss dataservices. if any exist, verifies/corrects to ensure they have the PC bit.
* checks the extension for zis plugins. if any exist, 
    * copies them to apf auth dataset
    * updates the zis parmlib

3) unless `--skip-enable` is given, the user YAML is updated to add 
```
components:
  <component-name>:
    enabled: true
```

The original description focused upon a subset of number 2 only.
1 & 3 can be done individually today while skipping 2. You just run them directly instead of  `zwe components install`. This would then require that the user take all actions pertaining to proclib, zis dataset, and parmlib updates on their own and update the YAML on their own. The actions of 2 can be explained in more detail for that, as well as improving its `--dry-run` to 
facilitate manual action. 

Currently `zwe components install` has `--dry-run` but the subcommands do not since nobody has ever been instructed to use them until now.

**Comment by nosrednayduj** [comment]
OK, that's cool, I had no idea about the subcommands.  So maybe I can ALREADY do what the customer wants by simply calling zwe components install extract always and process-hook only if the customer says they want the direct usage.  Which we'll need a  yaml bit for, and we need to do that design.

**Comment by Gautham-coder** [comment]
We are waiting on user feedback to decide on how to proceed with this.

**Comment by nosrednayduj** [comment]
Actually, we are waiting on us to make a proposal to the user that's more concrete. Maybe that means we're waiting on me. Maybe I'm the user in this case? Because of being the extension writer? Anyway, the user who pays money to rocket is not what were waiting on. We're waiting on rocket, which really means me.

**Comment by nosrednayduj** [comment]
I decided to try to out the subcommands, and realized that what I need is the ability to create a temporary file with the modified ZWESIP00 so that the customer can copy this into their parmlib. Maybe that's the dry run option that is needed for the subcommands. Because, we have an install.sh script which only runs configmgr env and sources the environment produced. Magically this environment is carried over to the rest of the call after you "dot" install.sh, because our variable templates that go into ZWESIP00 have references to $ZWE_components_izp_blah_blah and $ZWE_zowe_setup_blah_blah. So we need some procedure which goes through the ZIS plug-ins and does the work of determining whether or not each variable in the zisServer/samplib file is present in ZWESIP00, and either updates it if present, or adds it at the end if not, producing a temporary file, and simply avoiding copying it. Because it's really only the copy to parmlib that they don't want. (As well as the copy to authPluginLib.) We don't want to have to document going through 10 zisServer/samplib directories and manually updating ZWESIP00. We just want to document copying one file.

Also, you mentioned the subcommands were for zwe component install; would everything apply as well to zwe components upgrade?

upgrade --dry-run --trace doesn't print much.  What should it do?  Oh!  install --dry-run prints an error Invalid parameter --dry-run. So maybe it really doesn't do anything in any case.

**Comment by nosrednayduj** [comment]
I hate auto close.  Don't do it.  Just leave old things open!!
